### PR TITLE
Plugin [1700] - Http_enhancements

### PR DESCRIPTION
**Pull request by Shubhangi-cs**
Issue: HTTP Proxy Configuration Missing in HTTP Sink Plugin
**Fix **
To address this issue and provide users with HTTP Proxy support in the HTTP Sink plugin, we propose configuring the following fields:
HTTP Proxy :
Proxy URL
Proxy Username
Proxy Password
https://cdap.atlassian.net/browse/PLUGIN-1700

### CODE REVIEW

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 1]
```diff
@@ -65,6 +69,8 @@ public abstract class BaseHttpConfig extends ReferencePluginConfig {
 
     public static final String PROPERTY_SERVICE_ACCOUNT_SCOPE = "serviceAccountScope";
 
+    public static final String REGEX_PROXY_URL = "^(?i)(https?)://.*$";
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: Why are we making this change ?

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java [hunk 2]
```diff
@@ -354,6 +396,12 @@ public void validate(FailureCollector failureCollector) {
         }
     }
 
+    public void validateProxy() {
+        if (!getProxyUrl().matches(REGEX_PROXY_URL)) {
+            throw new IllegalArgumentException(String.format("Proxy URL format is wrong: %s.", getProxyUrl()));
+        }
+    }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: This does not look like a backward compatible change as there can be an older pipeline which might not follow this regex.

File: src/main/java/io/cdap/plugin/http/common/http/HttpClient.java [hunk 3]
```diff
@@ -167,10 +167,10 @@ private Header getAuthorizationHeader(AccessToken accessToken) {
   /**
    * This class allows us to send body not only in POST/PUT but also in other requests.
    */
-  private static class HttpRequest extends HttpEntityEnclosingRequestBase {
+  public static class HttpRequest extends HttpEntityEnclosingRequestBase {
     private final String methodName;
 
-    HttpRequest(URI uri, String methodName) {
+    public HttpRequest(URI uri, String methodName) {
       super();
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/common/http/HttpClient.java]: Move this class outside in a separate file.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 4]
```diff
@@ -158,54 +175,88 @@ private boolean executeHTTPServiceAndCheckStatusCode() throws IOException {
...
+            new AuthScope(proxyHost, proxyPort),
+            new UsernamePasswordCredentials(proxyUser, proxyPassword));
+        }
+
+        HttpHost proxy = new HttpHost(proxyHost, proxyPort);
+        httpClient = HttpClients.custom()
+          .setDefaultCredentialsProvider(credsProvider)
+          .setProxy(proxy)
+          .build();
+      }
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: Refarctor this portion of the code into a separate method, getHttpClient internally which you can either create HttpClients.createDefault or HttpClients.custom based on the use case.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 5]
```diff
@@ -245,7 +296,7 @@ private String updateURLWithPlaceholderValue(StructuredRecord inputRecord) {
   }
 
   private void flushMessageBuffer() {
-    if (messageBuffer.isEmpty()) {
+    if (messageBuffer.isEmpty() && !config.getMethod().equals("DELETE")) {
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: Please add comment about this change.

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 6]
```diff
@@ -158,54 +175,88 @@ private boolean executeHTTPServiceAndCheckStatusCode() throws IOException {
...
+      response = httpClient.execute(request);
+
+      httpStatusCode = response.getStatusLine().getStatusCode();
+      LOG.debug("Response HTTP Status code: {}", httpStatusCode);
+
     } catch (MalformedURLException | ProtocolException e) {
       throw new IllegalStateException("Error opening url connection. Reason: " + e.getMessage(), e);
     } catch (IOException e) {
-      LOG.warn("Error making {} request to url {} with headers {}.", config.getMethod(), config.getUrl(), headers);
+      LOG.warn("Error making {} request to url {}.", config.getMethod(), config.getUrl());
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java:210]: We should not log the header informations in the logs.

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java (hunk 1 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: Removed

File: src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java (hunk 2 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/http/common/BaseHttpConfig.java]: Removed

File: src/main/java/io/cdap/plugin/http/common/http/HttpClient.java (hunk 3 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/http/common/http/HttpClient.java]: Moved

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java (hunk 4 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: Done

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java (hunk 5 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]: Done

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java (hunk 6 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java:210]: Headers are removed in this change, previously it was there

File: src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java [hunk 7]
```diff
@@ -148,74 +172,137 @@ public boolean verify(String hostname, SSLSession session) {
...
+    }
+
+    // set default headers
+    if (headers != null) {
+      for (Map.Entry<String, String> headerEntry : this.headers.entrySet()) {
+        clientHeaders.add(new BasicHeader(headerEntry.getKey(), headerEntry.getValue()));
+      }
+    }
+
+    return clientHeaders.toArray(new Header[0]);
```
Comment (MrRahulSharma) [inline src/main/java/io/cdap/plugin/http/sink/batch/HTTPRecordWriter.java]:
```suggestion
    return clientHeaders.toArray(new Header[clientHeaders.size()]);
```

**Review by MrRahulSharma — APPROVED** [review]

**Pull request merged**